### PR TITLE
Sockets cq_sread should retry on -EINTR

### PR DESCRIPTION
**Pull request by chrisdolan**
Signed-off-by: Chris Dolan <chrisdolan@google.com>

### CODE REVIEW

File: prov/sockets/src/sock_cq.c [hunk 1]
```diff
@@ -406,7 +406,7 @@ static ssize_t sock_cq_sreadfrom(struct fid_cq *cq, void *buf, size_t count,
 		} while (ret > 0);
 	}
 
-	return (ret == 0 || ret == -FI_ETIMEDOUT) ? -FI_EAGAIN : ret;
+	return (ret == 0 || ret == -FI_ETIMEDOUT || ret == -EINTR) ? -FI_EAGAIN : ret;
```
Comment (shefty) [inline prov/sockets/src/sock_cq.c]: There was some discussion either around this or around a similar change way, way back.  @sydidelot may remember the details better that I can.

My general view is that we should not hide EINTR or other expected failures from the app.  Is there a reason to hide this?

Comment (chrisdolan) [reply prov/sockets/src/sock_cq.c]: Most of the client code I've seen doesn't check for -EINTR but only checks for -EAGAIN.
  OpenMPI: https://github.com/open-mpi/ompi/blob/master/ompi/mca/mtl/ofi/mtl_ofi.h#L227
  pingpong.c: https://github.com/ofiwg/libfabric/blob/master/util/pingpong.c#L1088
I can't remember how I encountered this exactly...

Comment (chrisdolan) [reply prov/sockets/src/sock_cq.c]: It's possible that I witnessed this when I suspended and resumed the process (perhaps ctrl-Z, fg; or maybe it was from attaching gdb) and it aborted with -EINTR. Again, my memory is fuzzy, but I think it was in this code path: https://github.com/open-mpi/ompi/blob/89c1b0c7edb173a2c94a3641767d0c86c0ad9021/orte/orted/orted_comm.c#L218

Comment (shefty) [reply prov/sockets/src/sock_cq.c]: I can believe that apps aren't checking for anything other than EAGAIN.  But the question is whether it's the responsibility of libfabric to convert other error codes to EAGAIN, or for the app to check other values.  Basically, what error codes _should_ OFI hide from the app?

Comment (shefty) [reply prov/sockets/src/sock_cq.c]: I posted a message to the ofiwg to gather community input, if any, on this PR.

**Comment by shefty** [comment]
I noticed this:

https://github.com/ofiwg/libfabric/blob/master/prov/rxm/src/rxm_conn.c#L1269

@a-ilango - I didn't easily find a comment on the EINTR check or why once gets set to 0.  I think there's a bug here, maybe a missing break from the loop when once = 0?  This only applies to EQ sread.  No other provider seem to handle EINTR for eq or cq sread calls.

I _think_ we want this change, but it needs to be coordinated across all providers and the man pages.  The sockets provider is being replaced by tcp+rxm, so additional changes are needed somewhere.

**Comment by chrisdolan** [comment]
I saw that, and interpreted the "once = 0" as only logging once per call instead of once per loop iteration.

**Comment by a-ilango** [comment]
I get EINTR errors when I'm debugging a program inside gdb. So this piece of code ignores those errors and prints the log just once to inform the user.

**Comment by shefty** [comment]
Hmm... I don't think we want to completely ignore EINTR.  The app may have a valid reason for signaling a blocked thread, so that it can return to the user.

**Comment by a-ilango** [comment]
It seems we would get a problem if we have a signal handler trapping SIGINT somewhere: http://250bpm.com/blog:12. That can lead to a case where the program doesn't end when the user issues ctrl-C.

Perhaps we should allow ignoring EINTR in debug builds as it's needed there.

**Comment by shefty** [comment]
Yes, we need to return on EINTR, but ignoring it in a debug build seems reasonable.

**Comment by shefty** [comment]
Merged this.  We should return on interrupt, so the rxm provider will need to be updated.  It's probably okay to ignore EINTR in a debug build if that's needed to make things work with a debugger.

**Pull request merged**